kong/pdk/vault.lua

```diff
@@ -307,728 +499,742 @@ local function new(self)
-      local key = keys[j]
-      local val = refs[key]
-      KEY_BUFFER:putf("%s=%s;", key, val)
+    if type(options) ~= "table" then
+      return references
     end
+
+    for key, value in pairs(options) do
+      if key ~= "$refs" and type(value) == "table" then
+        get_references(value, references, collected)
+      end
+    end
+
+    local refs = options["$refs"]
+    if type(refs) ~= "table" or isempty(refs) then
+      return references
+    end
+
+    for _, reference in pairs(refs) do
+      if type(reference) == "string" then -- a string reference
+        if not collected[reference] then
+          collected[reference] = true
+          collected.n = collected.n + 1
+          references[collected.n] = reference
+        end
+
+      elseif type(reference) == "table" then -- array, set or map of references
+        for _, ref in pairs(reference) do
+          if not collected[ref] then
+            collected[ref] = true
+            collected.n = collected.n + 1
+            references[collected.n] = ref
+          end
+        end
+      end
+    end
+
+    return references
   end
 
-  --- Generates an LRU (Least Recently Used) cache key based on sorted keys of the references.
-  -- This function generates a key for each reference, sorts these keys, and then populates a
-  -- key buffer with these keys. It also generates an md5 hash of the key buffer.
+
+  ---
+  -- Function `get_sorted_references` recursively iterates over options and returns
+  -- all the references in an sorted array. The same reference is in array only once.
+  --
   -- @local
-  -- @function populate_key_buffer
-  -- @tparam table refs The references based on which cache keys are to be generated.
-  -- @treturn table keys The sorted keys from the references.
-  -- @treturn number count The count of the keys.
-  -- @treturn string md5Hash The md5 hash of the populated key buffer.
-  -- @usage local keys, count, hash = populate_key_buffer(refs)
-  local function populate_key_buffer(refs)
-    -- Generate an LRU (Least Recently Used) cache key based on sorted keys of the references
-    local keys, count = generate_sorted_keys(refs)
-    populate_buffer(keys, count, refs)
-    return keys, count, md5_bin(KEY_BUFFER:get())
+  -- @function get_sorted_references
+  -- @tparam table options the options to look for the references
+  -- @treturn table|nil an sorted array of collected references, return `nil` in case no references were found.
+  --
+  -- @usage
+  -- local references = get_sorted_references({
+  --   username = "john",
+  --   password = "doe",
+  --   ["$refs"] = {
+  --     username = "{vault://aws/database/username}",
+  --     password = "{vault://aws/database/password}",
+  --   }
+  -- })
+  local function get_sorted_references(options)
+    local references = get_references(options)
+    if isempty(references) then
+      return
+    end
+
+    sort(references)
+
+    return references, encode_base64url(md5_bin(concat(references, ".")))
```

Comment:
The comment does not describe the second return value.   It would be clearer to calculate the hash at the call site anyway, I think.

Reply:
Ok. I move it to call site.